catalog/files/impl/src/main/java/org/projectnessie/catalog/files/s3/S3SessionsManager.java

```diff
@@ -212,6 +226,118 @@ private Credentials executeAssumeRoleRequest(
+      statement.put("Action", "s3:ListBucket");
+      statement.put("Resource", format("arn:aws:s3:::%s", bucket));
+      ObjectNode condition = statement.withObject("Condition");
+      ObjectNode stringLike = condition.withObject("StringLike");
+      ArrayNode s3Prefix = stringLike.withArray("s3:prefix");
+      s3Prefix.add(path);
+      s3Prefix.add(path + "/*");
+      // For write.object-storage.enabled=true
+      s3Prefix.add("*/" + path);
+      s3Prefix.add("*/" + path + "/*");
+    }
+
+    // "Allow Write" for all remaining S3 actions on the bucket+path.
+    List<StorageUri> writeable = locations.writeableLocations();
+    if (!writeable.isEmpty()) {
+      ObjectNode statement = statements.addObject();
+      statement.put("Effect", "Allow");
+      ArrayNode actions = statement.putArray("Action");
+      actions.add("s3:GetObject");
+      actions.add("s3:GetObjectVersion");
+      actions.add("s3:PutObject");
+      actions.add("s3:DeleteObject");
+      ArrayNode resources = statement.withArray("Resource");
+      for (StorageUri location : writeable) {
+        String bucket = iamEscapeString(location.requiredAuthority());
+        String path = iamEscapeString(location.pathWithoutLeadingTrailingSlash());
+        resources.add(format("arn:aws:s3:::%s/%s/*", bucket, path));
+        // For write.object-storage.enabled=true
+        resources.add(format("arn:aws:s3:::%s/*/%s/*", bucket, path));
+      }
+    }
+
+    // "Allow read" for all remaining S3 actions on the bucket+path.
+    List<StorageUri> readonly = locations.readonlyLocations();
+    if (!readonly.isEmpty()) {
+      ObjectNode statement = statements.addObject();
+      statement.put("Effect", "Allow");
+      ArrayNode actions = statement.putArray("Action");
+      actions.add("s3:GetObject");
+      actions.add("s3:GetObjectVersion");
+      ArrayNode resources = statement.withArray("Resource");
+      for (StorageUri location : readonly) {
+        String bucket = iamEscapeString(location.requiredAuthority());
+        String path = iamEscapeString(location.pathWithoutLeadingTrailingSlash());
+        resources.add(format("arn:aws:s3:::%s%s/*", bucket, path));
+        // For write.object-storage.enabled=true
+        resources.add(format("arn:aws:s3:::%s/*/%s/*", bucket, path));
+      }
+    }
+
+    // Add custom statements
+    clientIamStatements.ifPresent(
+        stmts -> stmts.stream().map(ParsedIamStatements.STATEMENTS::get).forEach(statements::add));
+
+    return policy.toString();
+  }
+
+  static class ParsedIamStatements {
+    static final LoadingCache<String, ObjectNode> STATEMENTS =
+        Caffeine.newBuilder()
+            .maximumSize(2000)
+            .expireAfterAccess(Duration.of(1, ChronoUnit.HOURS))
+            .build(ParsedIamStatements::parseStatement);
+
+    private static ObjectNode parseStatement(String stmt) {
+      try (MappingIterator<Object> values = MAPPER.readerFor(ObjectNode.class).readValues(stmt)) {
```

Comment:
Why `readValues` instead of `readValue`? Do we expect `stmt` to be empty or to contain many JSON objects?

Reply:
Necessary for configuration checks/validation.